src/SharedMemoryParcel.php

```diff
@@ -0,0 +1,445 @@
+    }
+
+    private function open(): void
+    {
+        $this->memOpen($this->key, 'w', 0, 0);
+    }
+
+    /**
+     * Checks if the object has been freed.
+     *
+     * Note that this does not check if the object has been destroyed; it only
+     * checks if this handle has freed its reference to the object.
+     *
+     * @return bool True if the object is freed, otherwise false.
+     */
+    private function isFreed(): bool
+    {
+        // If we are no longer connected to the memory segment, check if it has
+        // been invalidated.
+        if ($this->handle !== null) {
+            $this->handleMovedMemory();
+            $header = $this->getHeader();
+            return $header['state'] === self::STATE_FREED;
+        }
+
+        return true;
+    }
+
+    /**
+     * @return mixed
+     *
+     * @throws ParcelException
+     * @throws SerializationException
+     */
+    private function getValue(): mixed
+    {
+        if ($this->isFreed()) {
+            throw new ParcelException('The object has already been freed');
+        }
+
+        $header = $this->getHeader();
+
+        // Make sure the header is in a valid state and format.
+        if ($header['state'] !== self::STATE_ALLOCATED || $header['size'] <= 0) {
+            throw new ParcelException('Shared object memory is corrupt');
+        }
+
+        // Read the actual value data from memory and unserialize it.
+        $data = $this->memGet(self::MEM_DATA_OFFSET, $header['size']);
+        return $this->serializer->unserialize($data);
+    }
+
+    /**
+     * If the value requires more memory to store than currently allocated, a
+     * new shared memory segment will be allocated with a larger size to store
+     * the value in. The previous memory segment will be cleaned up and marked
+     * for deletion. Other processes and threads will be notified of the new
+     * memory segment on the next read attempt. Once all running processes and
+     * threads disconnect from the old segment, it will be freed by the OS.
+     */
+    private function wrap(mixed $value): void
+    {
+        if ($this->isFreed()) {
+            throw new ParcelException('The object has already been freed');
+        }
+
+        $serialized = $this->serializer->serialize($value);
+        $size = \strlen($serialized);
+        $header = $this->getHeader();
+
+        /* If we run out of space, we need to allocate a new shared memory
+           segment that is larger than the current one. To coordinate with other
+           processes, we will leave a message in the old segment that the segment
+           has moved and along with the new key. The old segment will be discarded
+           automatically after all other processes notice the change and close
+           the old handle.
+        */
+        if (\shmop_size($this->handle) < $size + self::MEM_DATA_OFFSET) {
+            $this->key = $this->key < 0xffffffff ? $this->key + 1 : \random_int(0x10, 0xfffffffe);
```

Comment:
What's happening here?

Reply:
Seems it is picking the next incremental ID or a random ID if the next one is too high? This seems very odd and either we should fail if the memory segment is too small or do something better.